src/reflect/scala/reflect/internal/Names.scala

```diff
@@ -14,607 +14,291 @@ package scala
-    private def paranoidEquals(other: Any): Boolean = {
-      val cmp = this eq other.asInstanceOf[AnyRef]
-      if (cmp || !nameDebug)
-        return cmp
-
-      other match {
-        case x: String  =>
-          Console.println(s"Compared $debugString and String '$x'")
-        case x: Name    =>
-          if (this.isTermName != x.isTermName) {
-            val panic = this.toTermName == x.toTermName
-            Console.println("Compared '%s' and '%s', one term, one type.%s".format(this, x,
-              if (panic) "  And they contain the same name string!"
-              else ""
-            ))
-          }
-        case _ =>
+    res
+    //same as cache.computeIfAbsent(value, new NameHolder(_)) but faster
+  }
+  private[this] final val cache = new ConcurrentHashMap[String, NameHolder](100000, 0.75F, 1)
+  //deprecated stuff
+  @deprecated @inline final def newTermNameCached(s: String): TermName = newTermName(s)
+  @deprecated @inline final def newTypeNameCached(s: String): TypeName = newTypeName(s)
+
+  class NameHolder(private[NameHolder] val value: String) {
+
+   lazy val decodedHolder: NameHolder = {
+      if (value contains '$') {
+        val res = NameTransformer.decode(value)
+        if (res == value) this
+        else newName(res)
       }
-      false
+      else this
     }
-    ****/
-
-    /** @return the i'th Char of this name */
-    final def charAt(i: Int): Char = chrs(index + i)
-
-    /** @return the index of first occurrence of char c in this name, length if not found */
-    final def pos(c: Char): Int = pos(c, 0)
-
-    /** @return the index of first occurrence of s in this name, length if not found */
-    final def pos(s: String): Int = pos(s, 0)
-
-    /** Returns the index of the first occurrence of character c in
-     *  this name from start, length if not found.
-     *
-     *  @param c     the character
-     *  @param start the index from which to search
-     *  @return      the index of the first occurrence of c
-     */
-    final def pos(c: Char, start: Int): Int = {
-      var i = start
-      while (i < len && chrs(index + i) != c) i += 1
-      i
+    lazy val encodedHolder: NameHolder = {
+      val res = NameTransformer.encode(value)
+      if (res == value) this else newName(res)
     }
 
-    /** Returns the index of the first occurrence of nonempty string s
-     *  in this name from start, length if not found.
-     *
-     *  @param s     the string
-     *  @param start the index from which to search
-     *  @return      the index of the first occurrence of s
-     */
-    final def pos(s: String, start: Int): Int = {
-      var i = pos(s.charAt(0), start)
-      val sLen = s.length()
-      if (sLen == 1) return i
-      while (i + sLen <= len) {
-        var j = 1
-        while (s.charAt(j) == chrs(index + i + j)) {
-          j += 1
-          if (j == sLen) return i
-        }
-        i = pos(s.charAt(0), i + 1)
+    abstract class AName extends NameApi with CharSequence {
```

Comment:
I worry slightly about each name string becoming 3 objects now; I'm not sure how common it is to convert between term and type names, now that `newTypeName` isn't implemented in terms of `newTermName`.